modules/frontend/combiner/response_metrics.go

```diff
@@ -0,0 +1,95 @@
+package combiner
+
+import (
+	"github.com/grafana/tempo/pkg/tempopb"
+)
+
+// These structs combine response metrics in a single place
+
+type SearchMetricsCombiner struct {
+	Metrics *tempopb.SearchMetrics
+}
+
+func NewSearchMetricsCombiner() *SearchMetricsCombiner {
+	return &SearchMetricsCombiner{
+		Metrics: &tempopb.SearchMetrics{},
+	}
+}
+
+func (mc *SearchMetricsCombiner) Combine(newMetrics *tempopb.SearchMetrics, resp PipelineResponse) {
+	if newMetrics != nil && !IsCacheHit(resp.HTTPResponse()) {
+		mc.Metrics.InspectedTraces += newMetrics.InspectedTraces
+		mc.Metrics.InspectedBytes += newMetrics.InspectedBytes
+		mc.Metrics.CompletedJobs++
+	}
+}
+
+func (mc *SearchMetricsCombiner) CombineMetadata(newMetrics *tempopb.SearchMetrics, resp PipelineResponse) {
+	if newMetrics != nil && !IsCacheHit(resp.HTTPResponse()) {
+		mc.Metrics.TotalBlocks += newMetrics.TotalBlocks
+		mc.Metrics.TotalJobs += newMetrics.TotalJobs
+		mc.Metrics.TotalBlockBytes += newMetrics.TotalBlockBytes
+	}
+}
+
+type TraceByIDMetricsCombiner struct {
+	Metrics *tempopb.TraceByIDMetrics
+}
+
+func NewTraceByIDMetricsCombiner() *TraceByIDMetricsCombiner {
+	return &TraceByIDMetricsCombiner{
+		Metrics: &tempopb.TraceByIDMetrics{},
+	}
+}
+
+func (mc *TraceByIDMetricsCombiner) Combine(newMetrics *tempopb.TraceByIDMetrics, resp PipelineResponse) {
+	if newMetrics != nil && !IsCacheHit(resp.HTTPResponse()) {
+		mc.Metrics.InspectedBytes += newMetrics.InspectedBytes
+	}
+}
+
+type MetadataMetricsCombiner struct {
+	Metrics *tempopb.MetadataMetrics
+}
+
+func NewMetadataMetricsCombiner() *MetadataMetricsCombiner {
+	return &MetadataMetricsCombiner{
+		Metrics: &tempopb.MetadataMetrics{},
+	}
+}
+
+func (mc *MetadataMetricsCombiner) Combine(newMetrics *tempopb.MetadataMetrics, resp PipelineResponse) {
+	if newMetrics != nil && !IsCacheHit(resp.HTTPResponse()) {
+		mc.Metrics.InspectedBytes += newMetrics.InspectedBytes
+	}
+}
+
+type QueryRangeMetricsCombiner struct {
+	Metrics *tempopb.SearchMetrics
+}
+
+func NewQueryRangeMetricsCombiner() *QueryRangeMetricsCombiner {
+	return &QueryRangeMetricsCombiner{
+		Metrics: &tempopb.SearchMetrics{},
+	}
+}
+
+func (mc *QueryRangeMetricsCombiner) Combine(newMetrics *tempopb.SearchMetrics, resp PipelineResponse) {
+	if newMetrics != nil {
+		// this is a coordination between the sharder and combiner. the sharder returns one response with summary metrics
```

Comment:
this could be removed using the same "meta data" pattern that the search pipeline uses. perhaps something you can tackle in a future PR?